### PR TITLE
[client] add new module `go.opentelemetry.io/collector/client`

### LINKED ISSUE #9804: Move client to a go module
The https://github.com/open-telemetry/opentelemetry-collector/tree/main/client folder should become a module so it can be delivered as a library and stabilized.

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Create a new go module `go.opentelemetry.io/collector/client`

**Link to tracking Issue:**
Fixes #9804

### CODE REVIEW

File: .chloggen/client_module.yaml
Comment (astencel-sumo) [inline .chloggen/client_module.yaml:25]:
```suggestion
change_logs: [api]
```

**Comment by bogdandrutu** [comment]
Please consider a big refactor (including renaming probably) for this package before marking stable.

**Comment by atoulme** [comment]
Thanks Bogdan, captured this feedback in https://github.com/open-telemetry/opentelemetry-collector/issues/9818

File: cmd/otelcorecol/go.mod
Comment (astencel-sumo) [inline cmd/otelcorecol/go.mod:5]: Not sure about this change, hope it will not trip the CI.

Comment (atoulme) [reply cmd/otelcorecol/go.mod:5]: weird, will revert

Comment (atoulme) [reply cmd/otelcorecol/go.mod:5]:
```suggestion
go 1.21
```

File: config/confighttp/go.mod
Comment (atoulme) [inline config/confighttp/go.mod:3]:
```suggestion
go 1.21
```

File: exporter/otlpexporter/go.mod
Comment (atoulme) [inline exporter/otlpexporter/go.mod:3]:
```suggestion
go 1.21
```

File: exporter/otlphttpexporter/go.mod
Comment (atoulme) [inline exporter/otlphttpexporter/go.mod:3]:
```suggestion
go 1.21
```

File: internal/e2e/go.mod
Comment (atoulme) [inline internal/e2e/go.mod:3]:
```suggestion
go 1.21
```

File: processor/batchprocessor/go.mod
Comment (atoulme) [inline processor/batchprocessor/go.mod:3]:
```suggestion
go 1.21
```

File: client/go.mod
Comment (atoulme) [inline client/go.mod:3]:
```suggestion
go 1.21
```

**Review by astencel-sumo — APPROVED** [review]

**Comment by atoulme** [comment]
There might be a few other ways to go about this ; closing for now this PR and let's maybe work on a refactoring approach.

File: client/go.mod
Comment (TylerHelmuth) [inline client/go.mod:1]: I'm ok with this name, but we could change it to `conn` or `connection` as proposed in https://github.com/open-telemetry/opentelemetry-collector/pull/9996

File: .github/workflows/build-and-test.yml
Comment (TylerHelmuth) [inline .github/workflows/build-and-test.yml:132]: Do we need this extra diff?

**Review by TylerHelmuth — COMMENTED** [review]
The Makefile's `check-contrib` and `restore-contrib` need updated with the new module.

File: client/go.mod
Comment (jpkrohling) [reply client/go.mod:1]: I very much prefer `client`, even if it's not perfect. I prefer this name, which has been in use for quite some time, rather than a breaking change only because of its name.

**Review by jpkrohling — APPROVED** [review]

File: client/go.mod
Comment (atoulme) [reply client/go.mod]: @jpkrohling this is not a small problem though - it breaks the build in a way that is not fixable right now, afaict.

**Review by mx-psi — APPROVED** [review]
I am in favor of `client` as a name. It's not perfect, but I don't have a better alternative in mind, and we have had no user complaints about it ever

**Review by TylerHelmuth — APPROVED** [review]

**Comment by mx-psi** [comment]
@atoulme can you fix the merge conflict?

@open-telemetry/collector-approvers I will merge this on Friday if nobody blocks before then

File: .chloggen/client_module.yaml
Comment (songy23) [inline .chloggen/client_module.yaml:18]:
```suggestion
subtext: This module contains generic representations of clients connecting to different receivers.
```

I feel it's worth adding a bit detail here for the changelog readers

**Review by songy23 — APPROVED** [review]

File: client/go.mod
Comment (codeboten) [inline client/go.mod:7]: the dependencies here need updating

File: config/confighttp/confighttp.go
Comment (codeboten) [inline config/confighttp/confighttp.go:144]: why is this change here?

File: client/go.mod
Comment (atoulme) [reply client/go.mod:7]:
```suggestion
	go.opentelemetry.io/collector/consumer v0.105.0
```

File: config/confighttp/confighttp.go
Comment (atoulme) [inline config/confighttp/confighttp.go:144]:
```suggestion
```

Comment (atoulme) [reply config/confighttp/confighttp.go]: fixed

**Review by codeboten — APPROVED** [review]

**Pull request merged**
